File: python/pomodoro.py
# -*- coding: utf-8 -*-
import os
import pickle
import time

from hashids import Hashids

from handlers import ProgressBarOutputHandler

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(SCRIPT_DIR, 'data')

STATUS_WORK = 'work'
STATUS_SHORT_REST = 'short'
STATUS_LONG_REST = 'long'


class PomodoroTimerNotFound(Exception):
    pass


def get_filename(obj_id):
    return os.path.join(DATA_DIR, obj_id + '.obj')


class PomodoroTimer(object):
    def __init__(self, work_duration=4, short_rest=2, long_rest=3,
                 output_handler_cls=ProgressBarOutputHandler,
                 id=None):
        self.id = self.__generate_id() if not id else id
        self.durations = {
            STATUS_WORK: work_duration,
            STATUS_SHORT_REST: short_rest,
            STATUS_LONG_REST: long_rest
        }
        self.output_handler = output_handler_cls(STATUS_WORK,
                                                 duration=work_duration)
        self.session_count = 0
        self.status = STATUS_WORK

    def __generate_id(self):
        hashids = Hashids(salt='T*{VghkFo}2lcknb~Np$s8+$^|n~>59~#,+D9Z3Eaq/n+',
                          alphabet='1234567890ABCDEF', min_length=6)
        return hashids.encrypt(int(time.time()))

    def dumps(self):
        filename = get_filename(self.id)
        f = open(filename, 'w')
        pickle.dump(self, f)
        f.close()

    def start(self):
        counter = self._counter

        while counter > 0:
            self.ticking(counter)
            counter -= 1
            time.sleep(1)

        self.timer_completed()
        self._next_status()

    def timer_completed(self):
        if self.is_working:
            self.session_count += 1
        self.output_handler.completed(self.session_count)

    def ticking(self, counter):
        self.output_handler.ticking(self.status, counter, self._counter)

    @property
    def is_working(self):
        return self.status == STATUS_WORK

    @property
    def is_short_rest(self):
        return self.status == STATUS_SHORT_REST

    @property
    def is_long_rest(self):
        return self.status == STATUS_LONG_REST

    def _next_status(self):
        if self.status == STATUS_WORK:
            if self.session_count % 4 == 0:
                self.status = STATUS_LONG_REST
            else:
                self.status = STATUS_SHORT_REST
        else:
            self.status = STATUS_WORK

        return self.status

    @property
    def _counter(self):
        return self.durations[self.status]

    @classmethod
    def loads(klass, id):
        if not id:
            raise PomodoroTimerNotFound('supply an id to load')
        try:
            filename = get_filename(id)
            f = open(filename, 'r')
            p = pickle.load(f)
            f.close()
            return p
        except IOError:
            raise PomodoroTimerNotFound(
                'Pomodoro timer not found: {}'.format(id))

File: ruby/pomodoro.rb
#!/usr/bin/env ruby

require 'securerandom'
require 'json'
require 'time'

STATUS_WORK = 'work'
STATUS_SHORT_REST = 'short'
STATUS_LONG_REST = 'long'

WORK_DURATION = 4
SHORT_REST_DURATION = 2
LONG_REST_DURATION = 3

SESSIONS_FILE = './sessions.json'

class PomodoroTimer
  attr_accessor :id, :status, :session_count, :durations

  def initialize(id: nil, work_duration: WORK_DURATION, short_rest: SHORT_REST_DURATION, long_rest: LONG_REST_DURATION)
    @id = id || generate_id
    @durations = {
      STATUS_WORK => work_duration,
      STATUS_SHORT_REST => short_rest,
      STATUS_LONG_REST => long_rest
    }
    @status = STATUS_WORK
    @session_count = 0
  end

  def start
    counter = current_duration

    while counter > 0
      ticking(counter)
      counter -= 1
      sleep(1)
    end

    timer_completed
    next_status
  end

  def save
    sessions = load_sessions
    sessions[@id] = {
      id: @id,
      status: @status,
      session_count: @session_count
    }
    File.write(SESSIONS_FILE, JSON.pretty_generate(sessions))
  end

  def self.load(id)
    sessions = load_sessions
    data = sessions[id]
    raise "PomodoroTimerNotFound: Timer with ID #{id} not found" unless data

    timer = new(id: data['id'])
    timer.status = data['status']
    timer.session_count = data['session_count']
    timer
  end

  private

  def generate_id
    SecureRandom.uuid
  end

  def current_duration
    @durations[@status]
  end

  def ticking(counter)
    puts "#{@status.capitalize} (#{@id}): #{counter} seconds remaining"
  end

  def timer_completed
    if working?
      @session_count += 1
    end
    puts "Completed #{@status} session (#{@id}). Total work sessions: #{@session_count}"
  end

  def next_status
    if working?
      @status = (@session_count % 4).zero? ? STATUS_LONG_REST : STATUS_SHORT_REST
    else
      @status = STATUS_WORK
    end
  end

  def working?
    @status == STATUS_WORK
  end

  def self.load_sessions
    if File.exist?(SESSIONS_FILE)
      JSON.parse(File.read(SESSIONS_FILE))
    else
      {}
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  id = ARGV[0]

  begin
    timer = PomodoroTimer.load(id)
    puts "Loaded existing timer with ID: #{id}"
  rescue
    timer = PomodoroTimer.new
    puts "No timer found with ID: #{id}. Created new timer with ID: #{timer.id}"
  end

  timer.start
  timer.save
end